### ISSUE TITLE
OpenGL requiring software doesn't work Ventura

### ISSUE THREAD

**Issue by debarshibanerjee** [XQuartz/XQuartz#365]
I am using MacOS Ventura 13.5.2 with Xquartz 2.8.5 on an M2 MacBook Pro.
There's a molecular visualization software called 'vmd' which I am running remotely (using ssh -X username@login.cluster), and it fails with the following error:

>Info) VMD for LINUXAMD64, version 1.9.4a57 (April 27, 2022)
Info) http://www.ks.uiuc.edu/Research/vmd/
Info) Email questions and bug reports to vmd@ks.uiuc.edu
Info) Please include this reference in published work using VMD:
Info)    Humphrey, W., Dalke, A. and Schulten, K., `VMD - Visual
Info)    Molecular Dynamics', J. Molec. Graphics 1996, 14.1, 33-38.
Info) -------------------------------------------------------------
Info) Multithreading available, 128 CPUs.
Info)   CPU features: SSE2 SSE4.1 AVX AVX2 FMA F16 AVX512F AVX512CD HT
Info) Free system memory: 238GB (94%)
Info) Creating CUDA device pool and initializing hardware...
Info) Detected 2 available CUDA accelerators, 1 P2P link, 1 island:
Info) [0-1] NVIDIA RTX A6000     84 SM_8.6 1.8 GHz, 48GB RAM SP32 KT AE2 ZC
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
ERROR) Could not create OpenGL rendering context-> Exiting...
Info) VMD for LINUXAMD64, version 1.9.4a57 (April 27, 2022)
Info) Unable to create OpenGL window.

Can you help me debug this issue? Thanks in advance!

**Comment by benz0li** [comment]
Most likely due to issue https://github.com/XQuartz/XQuartz/issues/335.

**Comment by benz0li** [comment]
Same for `monteverdi` (ORFEO Toolbox image processing GUI) rom my [Data Science Dev Containers](https://github.com/b-data/data-science-devcontainers) (R qgisprocess image):

<img width="812" alt="monteverdi critical error" src="https://github.com/XQuartz/XQuartz/assets/20233085/00735872-b606-42d1-9973-12396e5489ad">

```
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-root'
WARNG> Failed to restore window layout!
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
WARNG> QOpenGLWidget: Failed to create context
WARNG> QOpenGLWidget: Failed to create context
WARNG> QOpenGLWidget: Failed to create context
WARNG> QOpenGLWidget: Failed to create context
WARNG> Required OpenGL version '3.0.0' with GLSL version '1.30'.
WARNG> composeAndFlush: QOpenGLContext creation failed
WARNG> composeAndFlush: makeCurrent() failed
WARNG> composeAndFlush: makeCurrent() failed
WARNG> Current OpenGL version is '' supporting OpenGL Shading-Language (GLSL) version ''.
To run at best performances, this application needs, at least, OpenGL version '3.0.0' with GLSL version '1.30'.
The application will automatically switch to a rendering mode which does not make use of OpenGL shaders and GLSL.
If you are running this application under some remote-desktop service, runtime OpenGL and GLSL versions may differ from those running directly on remote platform.
...
```

**Comment by debarshibanerjee** [comment]
Do you know if there's any way around this? Are there any alternatives?

**Comment by benz0li** [comment]
> Do you know if there's any way around this?

Not with XQuartz.

---

> Are there any alternatives?

@debarshibanerjee Yes. Use a Linux VM instead:

1. [Download](https://docs.getutm.app/installation/macos/) and install [UTM](https://github.com/utmapp/UTM).
1. Download an `arm64` installation image (ISO) for a Linux distribution of your choice.  
   (My preference: Debian)
1. Create a new VM:
    * Start: Select Virtualize
    * Operating System: Select Linux
    * Linux: Boot ISO Image > Browse...: Open ISO file
    * Hardware
        * Set Memory
        * Set number of CPU cores
        * Enable Hardware OpenGL acceleration
    * Storage: Set size
    * (Optional) Shared Directory
    * Summary: Name your VM
1. Start VM and work through the installation process.
    * At 'Software selection': Select a desktop environment of your choice.
      (My preference: Xfce; without selecting 'Debian desktop environment')
    * At 'Finish the installation': Clear the VM's CD/DVD in UTM. Then, click continue.

---

[UTM supports HiDPI/Retina displays](https://docs.getutm.app/settings-qemu/devices/display/#retina-mode). E.g. for standard 2160p/4K UHD (3840x2160):

1. Stop VM.
1. UTM: Enable 'Retina Mode'.
1. Start VM.
1. VM: Change the following settings:
    * Display > General: Resolution: 3840x2160
    * Appearance > Settings: Window Scaling: 2x
    * Window Manager > Style: Theme: Default-xhdpi
    * Mouse and Touchpad > Theme: Size: Increase/Decrease 'Cursor size' if required.

ℹ️ I do not use the [SPICE Agent](https://docs.getutm.app/guest-support/linux/#spice-agent) for [Dynamic Resolution](https://docs.getutm.app/guest-support/dynamic-resolution/). I simply work in full screen at native resolution.

---

(**Xfce specific**) Should you require a custom resolution e.g. for a 16" MacBook Pro (2023):

1. Create file `~/.config/xfce4/xinitrc` with the following content:
   ```
   #!/bin/sh
   
   [ -f "$HOME/.xprofile" ] && . "$HOME/.xprofile"
   . /etc/xdg/xfce4/xinitrc
   ```
1. Create file `~/.xprofile` with the following content:  
   ```
   xrandr --newmode "3456x2160_60.00"  641.53  3456 3736 4120 4784  2160 2161 2164 2235  -HSync +Vsync
   xrandr --addmode Virtual-1 "3456x2160_60.00"
   xrandr -s 3456x2160
   ```  
   :information_source: Values according to command `gtf 3456 2160 60`[^1]
1. Log out.
1. Log in.

[^1]: 2160 = 2234 - height of notch. There is an additional strip (height: 10 px) which you could subtract...

(Comment out the last line in `~/.xprofile`, if you do not want the custom resolution being set automatically at login)

---

I recommend setting up a [shared directory](https://docs.getutm.app/guest-support/sharing/directory/) using [VirtFS](https://docs.getutm.app/guest-support/linux/#virtfs) (UTM) + `bindfs` (VM).  
ℹ️ There is no need for the [SPICE WebDAV](https://docs.getutm.app/guest-support/linux/#spice-webdav).

---

If you `ssh -X` into your remote server from the Linux VM, everything *should* work as expected.

**Comment by debarshibanerjee** [comment]
@benz0li Thanks, I followed your guide and it works quite well!

**Comment by benz0li** [comment]
@debarshibanerjee The instructions in https://github.com/XQuartz/XQuartz/issues/365#issuecomment-1762918977 have been updated. The setup now includes the option of setting the display's DPI (default: 96).

This ensures that Qt applications (and maybe others[^1]) scale up automatically. Should a certain application still not scale up, see https://wiki.archlinux.org/title/HiDPI#Applications for further instructions.

[^1]: XTerm for example

**Comment by jeremyhu** [comment]
Correct.  IGLX is not really supported any more.  It wasn't really supported well even when it was supported...

**Comment by xlla** [comment]
> If you `ssh -X` into your remote server from the Linux VM, everything _should_ work as expected.

the UTM use backend QEMU, the QEMU on macos is use software (cpu) rendering without some patch. 
so forward some GPU rendering from remote server into Linux VM, then rendering by hypervisor and host cpu is much lag.

I think recompile mesa with x11 enabled, it will enable llvmpipe and brings OpenGL 4.5,  then install xorg-server and lauch ours X11.app, then run
`ssh -X` into your remote server 
it will faster than VM approach.

the best method is someone and patch and fix mesa on macos to provide high version OpenGL implement in direct render mode.

**Comment by benz0li** [comment]
> the UTM use backend QEMU, the QEMU on macos is use software (cpu) rendering without some patch.

Display settings in UTM:

<img width="799" alt="display" src="https://github.com/XQuartz/XQuartz/assets/20233085/7d169d06-f261-4521-8fa3-f0fe555aacf5">

"About Xfce" in VM (Debian 12):

<img width="560" alt="virgl" src="https://github.com/XQuartz/XQuartz/assets/20233085/47ab1c1d-d8ca-4473-bb43-c57888900c39">

> VirGL is a virtual 3D GPU for use inside QEMU virtual machines, that allows the guest operating system to use the capabilities of the host GPU to accelerate 3D rendering.

– [VirGL &#8212; The Mesa 3D Graphics Library latest documentation](https://docs.mesa3d.org/drivers/virgl.html)

**Comment by Gbrothers1** [comment]
You can skip the VM work and just use VirtualGL over X11 Forwarding via SSH, works well for me on the LAN and on macOS over xQuartz, there is a dmg release on their GitHub for macOS and I can confirm it does work to load GLX on non-glx native drivers Apple M1 Metal. They no longer support openGL in favor of Metal. But you can still run it this way, I use this method of Tailscale to easily exposed IP's in my VPN. Currently using this technique to run Genesis AI on and RTX-GPU enable remote server on my macOS xQuartz vlgclient renderer.
